File: ios/ui-framework-sample/UIFramework/Sources/UIFramework/BizTransferList.swift
import SwiftUI

public struct BizTransferOption: Identifiable, Equatable {
    public let id: String
    public var key: String
    public var label: String
    public var disabled: Bool
    
    public init(key: String, label: String, disabled: Bool = false) {
        self.id = key
        self.key = key
        self.label = label
        self.disabled = disabled
    }
}

public struct BizTransferList: View {
    @Binding var modelValue: [String]
    var data: [BizTransferOption]
    var titles: [String]
    var variant: Variant
    var disabled: Bool
    var onChange: (([String], String, [String]) -> Void)?
    
    public enum Variant {
        case outlined
        case filled
        case underlined
    }
    
    @State private var leftChecked: Set<String> = []
    @State private var rightChecked: Set<String> = []
    
    public init(
        modelValue: Binding<[String]>,
        data: [BizTransferOption],
        titles: [String] = ["Source", "Target"],
        variant: Variant = .outlined,
        disabled: Bool = false,
        onChange: (([String], String, [String]) -> Void)? = nil
    ) {
        self._modelValue = modelValue
        self.data = data
        self.titles = titles
        self.variant = variant
        self.disabled = disabled
        self.onChange = onChange
    }
    
    public var body: some View {
        HStack(alignment: .center, spacing: 16) {
            // Left Panel
            transferPanel(
                title: titles[0],
                items: leftData,
                checkedSet: $leftChecked
            )
            
            // Actions
            VStack(spacing: 12) {
                Button(action: moveToRight) {
                    Image(systemName: "chevron.right")
                        .font(.system(size: 16, weight: .semibold))
                        .foregroundColor(
                            (disabled || leftChecked.isEmpty) ? Color(hex: "#CCCCCC") : Color(hex: "#333333")
                        )
                        .frame(width: 36, height: 36)
                        .background((disabled || leftChecked.isEmpty) ? Color(hex: "#F5F5F5") : Color.white)
                        .cornerRadius(4)
                        .overlay(
                            RoundedRectangle(cornerRadius: 4)
                                .stroke((disabled || leftChecked.isEmpty) ? Color(hex: "#EEEEEE") : Color(hex: "#CCCCCC"), lineWidth: 1)
                        )
                }
                .disabled(disabled || leftChecked.isEmpty)
                
                Button(action: moveToLeft) {
                    Image(systemName: "chevron.left")
                        .font(.system(size: 16, weight: .semibold))
                        .foregroundColor(
                            (disabled || rightChecked.isEmpty) ? Color(hex: "#CCCCCC") : Color(hex: "#333333")
                        )
                        .frame(width: 36, height: 36)
                        .background((disabled || rightChecked.isEmpty) ? Color(hex: "#F5F5F5") : Color.white)
                        .cornerRadius(4)
                        .overlay(
                            RoundedRectangle(cornerRadius: 4)
                                .stroke((disabled || rightChecked.isEmpty) ? Color(hex: "#EEEEEE") : Color(hex: "#CCCCCC"), lineWidth: 1)
                        )
                }
                .disabled(disabled || rightChecked.isEmpty)
            }
            
            // Right Panel
            transferPanel(
                title: titles.count > 1 ? titles[1] : "Target",
                items: rightData,
                checkedSet: $rightChecked
            )
        }
        .opacity(disabled ? 0.6 : 1.0)
    }
    
    private var leftData: [BizTransferOption] {
        data.filter { !modelValue.contains($0.key) }
    }
    
    private var rightData: [BizTransferOption] {
        data.filter { modelValue.contains($0.key) }
    }
    
    private func transferPanel(
        title: String,
        items: [BizTransferOption],
        checkedSet: Binding<Set<String>>
    ) -> some View {
        VStack(spacing: 0) {
            // Header
            HStack {
                Button(action: {
                    let allEnabled = items.filter { !$0.disabled }
                    if checkedSet.wrappedValue.count > 0 && checkedSet.wrappedValue.count == allEnabled.count {
                        checkedSet.wrappedValue.removeAll()
                    } else {
                        checkedSet.wrappedValue = Set(allEnabled.map { $0.key })
                    }
                }) {
                    HStack(spacing: 8) {
                        Image(systemName: (checkedSet.wrappedValue.count > 0 && checkedSet.wrappedValue.count == items.filter { !$0.disabled }.count) ? "checkmark.square.fill" : "square")
                            .foregroundColor(
                                (checkedSet.wrappedValue.count > 0 && checkedSet.wrappedValue.count == items.filter { !$0.disabled }.count) ? Color(hex: "#1A2A5E") : Color(hex: "#CCCCCC")
                            )
                        
                        Text(title)
                            .font(.system(size: 14, weight: .medium))
                            .foregroundColor(Color(hex: "#333333"))
                    }
                }
                .disabled(disabled)
                
                Spacer()
                
                Text("\(checkedSet.wrappedValue.count)/\(items.count)")
                    .font(.system(size: 12))
                    .foregroundColor(Color(hex: "#999999"))
            }
            .padding(.horizontal, variant == .underlined ? 8 : 16)
            .padding(.vertical, 12)
            .background(headerBackground)
            .overlay(
                Rectangle().frame(height: 1).foregroundColor(headerBorder),
                alignment: .bottom
            )
            
            // Body
            ScrollView {
                if items.isEmpty {
                    Text("No Data")
                        .font(.system(size: 14))
                        .foregroundColor(Color(hex: "#CCCCCC"))
                        .frame(maxWidth: .infinity, maxHeight: .infinity)
                        .padding(.top, 40)
                } else {
                    LazyVStack(spacing: 0) {
                        ForEach(items) { item in
                            Button(action: {
                                if !disabled && !item.disabled {
                                    if checkedSet.wrappedValue.contains(item.key) {
                                        checkedSet.wrappedValue.remove(item.key)
                                    } else {
                                        checkedSet.wrappedValue.insert(item.key)
                                    }
                                }
                            }) {
                                HStack(spacing: 8) {
                                    Image(systemName: checkedSet.wrappedValue.contains(item.key) ? "checkmark.square.fill" : "square")
                                        .foregroundColor(
                                            checkedSet.wrappedValue.contains(item.key) ? Color(hex: "#1A2A5E") : Color(hex: "#CCCCCC")
                                        )
                                    
                                    Text(item.label)
                                        .font(.system(size: 14))
                                        .foregroundColor(Color(hex: "#333333"))
                                    
                                    Spacer()
                                }
                                .padding(.horizontal, variant == .underlined ? 8 : 16)
                                .padding(.vertical, 8)
                                .background(checkedSet.wrappedValue.contains(item.key) ? Color(hex: "#1A2A5E").opacity(0.05) : Color.clear)
                            }
                            .buttonStyle(PlainButtonStyle())
                            .disabled(disabled || item.disabled)
                            .opacity(item.disabled ? 0.5 : 1.0)
                        }
                    }
                    .padding(.vertical, 8)
                }
            }
        }
        .frame(minWidth: 150, maxWidth: .infinity)
        .frame(height: 250)
        .background(panelBackground)
        .cornerRadius(variant == .outlined || variant == .filled ? 8 : 0)
        .overlay(panelBorder)
    }
    
    private var panelBackground: Color {
        switch variant {
        case .outlined: return Color.white
        case .filled: return Color(hex: "#F5F5F5")
        case .underlined: return Color.clear
        }
    }
    
    private var headerBackground: Color {
        switch variant {
        case .outlined: return Color(hex: "#FAFAFA")
        case .filled: return Color(hex: "#E8E8E8")
        case .underlined: return Color.clear
        }
    }
    
    private var headerBorder: Color {
        switch variant {
        case .outlined: return Color(hex: "#EEEEEE")
        case .filled: return Color(hex: "#DDDDDD")
        case .underlined: return Color.clear
        }
    }
    
    @ViewBuilder
    private var panelBorder: some View {
        switch variant {
        case .outlined:
            RoundedRectangle(cornerRadius: 8)
                .stroke(Color(hex: "#CCCCCC"), lineWidth: 1)
        case .filled:
            RoundedRectangle(cornerRadius: 8)
                .stroke(Color(hex: "#999999"), lineWidth: 0)
                .overlay(
                    Rectangle().frame(height: 2).foregroundColor(Color(hex: "#999999")),
                    alignment: .bottom
                )
        case .underlined:
            Rectangle()
                .frame(height: 2)
                .foregroundColor(Color(hex: "#CCCCCC"))
                .frame(maxHeight: .infinity, alignment: .bottom)
        }
    }
    
    private func moveToRight() {
        let moved = Array(leftChecked)
        modelValue.append(contentsOf: moved)
        onChange?(modelValue, "right", moved)
        leftChecked.removeAll()
    }
    
    private func moveToLeft() {
        let moved = Array(rightChecked)
        modelValue.removeAll { moved.contains($0) }
        onChange?(modelValue, "left", moved)
        rightChecked.removeAll()
    }
}

File: android/ui-framework-sample/ui-framework/src/main/java/kh/mcnc/lib/components/BizTransferList.kt
package kh.mcnc.lib.components

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.Checkbox
import androidx.compose.material3.Icon
import androidx.compose.material3.Text
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.alpha
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.vector.ImageVector
import androidx.compose.ui.graphics.vector.PathBuilder
import androidx.compose.ui.graphics.vector.path
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp

data class TransferOption(
    val key: String,
    val label: String,
    val disabled: Boolean = false
)

enum class BizTransferVariant {
    Outlined, Filled, Underlined
}

@Composable
fun BizTransferList(
    data: List<TransferOption>,
    targetKeys: List<String>,
    onTargetKeysChange: (List<String>) -> Unit,
    modifier: Modifier = Modifier,
    titles: Pair<String, String> = "Source" to "Target",
    variant: BizTransferVariant = BizTransferVariant.Outlined,
    disabled: Boolean = false
) {
    var leftChecked by remember { mutableStateOf(setOf<String>()) }
    var rightChecked by remember { mutableStateOf(setOf<String>()) }

    val leftData = data.filter { it.key !in targetKeys }
    val rightData = data.filter { it.key in targetKeys }

    fun moveToRight() {
        val newTarget = targetKeys + leftChecked
        onTargetKeysChange(newTarget)
        leftChecked = emptySet()
    }

    fun moveToLeft() {
        val newTarget = targetKeys.filter { it !in rightChecked }
        onTargetKeysChange(newTarget)
        rightChecked = emptySet()
    }

    Row(
        modifier = modifier
            .fillMaxWidth()
            .height(250.dp)
            .let { if (disabled) it.alpha(0.6f) else it },
        horizontalArrangement = Arrangement.spacedBy(16.dp),
        verticalAlignment = Alignment.CenterVertically
    ) {
        // Left Panel
        TransferPanel(
            title = titles.first,
            data = leftData,
            checkedKeys = leftChecked,
            onCheckedKeysChange = { leftChecked = it },
            variant = variant,
            disabled = disabled,
            modifier = Modifier.weight(1f).fillMaxHeight()
        )

        // Actions
        Column(
            verticalArrangement = Arrangement.spacedBy(12.dp),
            horizontalAlignment = Alignment.CenterHorizontally
        ) {
            TransferButton(
                isRight = true,
                onClick = { moveToRight() },
                enabled = !disabled && leftChecked.isNotEmpty()
            )
            TransferButton(
                isRight = false,
                onClick = { moveToLeft() },
                enabled = !disabled && rightChecked.isNotEmpty()
            )
        }

        // Right Panel
        TransferPanel(
            title = titles.second,
            data = rightData,
            checkedKeys = rightChecked,
            onCheckedKeysChange = { rightChecked = it },
            variant = variant,
            disabled = disabled,
            modifier = Modifier.weight(1f).fillMaxHeight()
        )
    }
}

@Composable
private fun TransferPanel(
    title: String,
    data: List<TransferOption>,
    checkedKeys: Set<String>,
    onCheckedKeysChange: (Set<String>) -> Unit,
    variant: BizTransferVariant,
    disabled: Boolean,
    modifier: Modifier
) {
    val allEnabledData = data.filter { !it.disabled }
    val isAllChecked = allEnabledData.isNotEmpty() && checkedKeys.containsAll(allEnabledData.map { it.key })
    val shape = if (variant == BizTransferVariant.Underlined) RoundedCornerShape(0.dp) else RoundedCornerShape(8.dp)

    val baseModifier = modifier.clip(shape)
    val panelModifier = when (variant) {
        BizTransferVariant.Outlined -> baseModifier.background(Color.White).border(1.dp, Color(0xFFCCCCCC), shape)
        BizTransferVariant.Filled -> baseModifier.background(Color(0xFFF5F5F5)) // Usually bottom border is drawn differently, omitting for simplicity
        BizTransferVariant.Underlined -> baseModifier.background(Color.Transparent)
    }

    Column(modifier = panelModifier) {
        // Header
        val headerBg = when (variant) {
            BizTransferVariant.Outlined -> Color(0xFFFAFAFA)
            BizTransferVariant.Filled -> Color(0xFFE8E8E8)
            BizTransferVariant.Underlined -> Color.Transparent
        }
        val headerPadding = if (variant == BizTransferVariant.Underlined) PaddingValues(horizontal = 8.dp, vertical = 12.dp)
        else PaddingValues(horizontal = 16.dp, vertical = 12.dp)

        Row(
            modifier = Modifier
                .fillMaxWidth()
                .background(headerBg)
                .padding(headerPadding),
            verticalAlignment = Alignment.CenterVertically,
            horizontalArrangement = Arrangement.SpaceBetween
        ) {
            Row(verticalAlignment = Alignment.CenterVertically) {
                Checkbox(
                    checked = isAllChecked,
                    onCheckedChange = { checked ->
                        if (checked) onCheckedKeysChange(allEnabledData.map { it.key }.toSet())
                        else onCheckedKeysChange(emptySet())
                    },
                    enabled = !disabled,
                    modifier = Modifier.size(24.dp)
                )
                Spacer(modifier = Modifier.width(8.dp))
                Text(text = title, fontSize = 14.sp, fontWeight = FontWeight.Medium, color = Color(0xFF333333))
            }
            Text(text = "${checkedKeys.size}/${data.size}", fontSize = 12.sp, color = Color(0xFF999999))
        }

        // Dividers
        if (variant == BizTransferVariant.Outlined) {
            Box(Modifier.fillMaxWidth().height(1.dp).background(Color(0xFFEEEEEE)))
        } else if (variant == BizTransferVariant.Filled) {
            Box(Modifier.fillMaxWidth().height(1.dp).background(Color(0xFFDDDDDD)))
        } else if (variant == BizTransferVariant.Underlined) {
            Box(Modifier.fillMaxWidth().height(2.dp).background(Color(0xFFCCCCCC)))
        }

        // Body
        if (data.isEmpty()) {
            Box(modifier = Modifier.fillMaxSize(), contentAlignment = Alignment.Center) {
                Text(text = "No Data", color = Color(0xFFCCCCCC), fontSize = 14.sp)
            }
        } else {
            LazyColumn(
                modifier = Modifier
                    .fillMaxSize()
                    .padding(vertical = 8.dp)
            ) {
                items(data, key = { it.key }) { item ->
                    val isChecked = checkedKeys.contains(item.key)
                    val itemDisabled = disabled || item.disabled
                    Row(
                        modifier = Modifier
                            .fillMaxWidth()
                            .clickable(enabled = !itemDisabled) {
                                if (isChecked) onCheckedKeysChange(checkedKeys - item.key)
                                else onCheckedKeysChange(checkedKeys + item.key)
                            }
                            .background(if (isChecked) Color(0x0D1A2A5E) else Color.Transparent)
                            .padding(
                                horizontal = if (variant == BizTransferVariant.Underlined) 8.dp else 16.dp,
                                vertical = 8.dp
                            ),
                        verticalAlignment = Alignment.CenterVertically
                    ) {
                        Checkbox(
                            checked = isChecked,
                            onCheckedChange = null,
                            enabled = !itemDisabled,
                            modifier = Modifier.size(24.dp)
                        )
                        Spacer(modifier = Modifier.width(8.dp))
                        Text(
                            text = item.label,
                            fontSize = 14.sp,
                            color = if (itemDisabled) Color(0xFF333333).copy(alpha = 0.5f) else Color(0xFF333333)
                        )
                    }
                }
            }
        }
    }
}

@Composable
private fun TransferButton(
    isRight: Boolean,
    onClick: () -> Unit,
    enabled: Boolean
) {
    Box(
        modifier = Modifier
            .size(36.dp)
            .clip(RoundedCornerShape(4.dp))
            .background(if (enabled) Color.White else Color(0xFFF5F5F5))
            .border(1.dp, if (enabled) Color(0xFFCCCCCC) else Color(0xFFEEEEEE), RoundedCornerShape(4.dp))
            .clickable(enabled = enabled, onClick = onClick),
        contentAlignment = Alignment.Center
    ) {
        val vector = androidx.compose.ui.graphics.vector.ImageVector.Builder(
            name = if (isRight) "Right" else "Left",
            defaultWidth = 18.dp,
            defaultHeight = 18.dp,
            viewportWidth = 24f,
            viewportHeight = 24f
        ).apply {
            path(
                stroke = androidx.compose.ui.graphics.SolidColor(if (enabled) Color(0xFF333333) else Color(0xFFCCCCCC)),
                strokeLineWidth = 2f,
                strokeLineCap = androidx.compose.ui.graphics.StrokeCap.Round,
                strokeLineJoin = androidx.compose.ui.graphics.StrokeJoin.Round
            ) {
                if (isRight) {
                    moveTo(9f, 18f)
                    lineTo(15f, 12f)
                    lineTo(9f, 6f)
                } else {
                    moveTo(15f, 18f)
                    lineTo(9f, 12f)
                    lineTo(15f, 6f)
                }
            }
        }.build()

        Icon(
            imageVector = vector,
            contentDescription = null,
            tint = Color.Unspecified,
            modifier = Modifier.size(18.dp)
        )
    }
}
